src/classes/CW_CampaignSitesTriggerHandler.cls

```diff
@@ -0,0 +1,109 @@
+global class CW_CampaignSitesTriggerHandler {
+/**
+* File Name   :    CW_CampaignSitesTriggerHandler.cls
+* Description :    
+*
+* Modification Log
+* ============================================================================
+* Ver Date       Author         Modification
+* --- ---------- -------------- --------------------------
+* 0.1 21/05/2014 Chris Yorskton Initial
+* 0.2 28/05/2014 Klea Kolaric	Fixing trigger chunking
+* 0.3 16/06/2014 Chris Yorskton Added the aggregation of Brands on Campaign object
+*
+*/	
+	global static Boolean futureCalled = false;	// avoid multiple calls to future method
+	
+	private Boolean m_isExecuting = false;
+	
+    global CW_CampaignSitesTriggerHandler(Boolean isExecuting){
+        m_isExecuting = isExecuting;
+    }
+    
+    global boolean IsTriggerContext{
+        get{ return m_isExecuting; }
+    }
+	
+	global void OnAfterInsert( CW_Campaign_Sites__c[] newObjects, Map<Id,CW_Campaign_Sites__c> newObjectMap ) {
+		callFuture(newObjects);
+    }
+    
+	global void OnAfterDelete( CW_Campaign_Sites__c[] oldObjects) {
+		callFuture(oldObjects);
+    }
+	
+	/*global void OnAfterUpdate(
+		CW_Campaign_Sites__c[] oldObjects, 
+		CW_Campaign_Sites__c[] updatedObjects, 
+		Map<Id,CW_Campaign_Sites__c> oldObjectMap, 
+		Map<Id,CW_Campaign_Sites__c> updatedObjectMap
+	) {
+		updateCampaignWithAggregateFields(oldObjects); //todo should this be oldObjects?
+	}*/
+	
+	global void callFuture(CW_Campaign_Sites__c[] csList)
+	{
+		if(futureCalled) return;
+		
+		Set<Id> campaignIds = new Set<Id>();
+		for(CW_Campaign_Sites__c cs : csList)
+		{
+			campaignIds.add(cs.Campaign__c);
+		}
+		updateCampaignWithAggregateFields(campaignIds);
+		
+		futureCalled = true;
+	}
+	
+	@future
+	global static void updateCampaignWithAggregateFields(Set<Id> campaignIds) {
+		
+		
+		// maps that will contain aggregated code values - using set to de-dup
+		// k: campaign.id; v: set of label/language/product codes
+		Map<Id,Set<String>> lablsMap = new Map<Id,Set<String>>();
+		Map<Id,Set<String>> langsMap = new Map<Id,Set<String>>();
+		Map<Id,Set<String>> prodsMap = new Map<Id,Set<String>>();
+		Map<Id,Set<String>> brandsMap = new Map<Id,Set<String>>();
+		
+		// collect all campaign asset ids
+		for(Id campaignId : campaignIds) {
+			lablsMap.put(campaignId,new Set<String>());
+			langsMap.put(campaignId,new Set<String>());
+			prodsMap.put(campaignId,new Set<String>());
+			brandsMap.put(campaignId,new Set<String>());
+		}
+		
+		// reselect campaign site based on passed in ids with all needed fields
+		for(CW_Campaign_Sites__c cs : [
+			select 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.